### PR TITLE
[SMALLFIX] Improve test logging

### PR DESCRIPTION
**Pull request by aaudiber**
This PR uses a RuleChain to enforce an order on rules, allowing us to put a log handling rule at the outermost level and have it capture logs generated by other rules like LocalAlluxioClusterResource.

@calvinjia @apc999 PTAL. If you're on board with the idea, I'll update all integration tests to pass their rules through the `rules()` method

### CODE REVIEW

**Comment by aaudiber** [comment]
Jenkins, test this please

File: minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java [hunk 1]
```diff
@@ -241,8 +243,8 @@ public synchronized void saveWorkdir() throws IOException {
     ProcessBuilder pb =
         new ProcessBuilder("tar", "-czf", tarball.getName(), mWorkDir.getName());
     pb.directory(mWorkDir.getParentFile());
-    pb.redirectOutput(TESTS_LOG);
-    pb.redirectError(TESTS_LOG);
+    pb.redirectOutput(Redirect.appendTo(TESTS_LOG));
```
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/multi/process/MultiProcessCluster.java]: default behavior is to overwrite the log

File: tests/src/test/java/alluxio/BaseIntegrationTest.java [hunk 2]
```diff
@@ -21,41 +23,65 @@
...
+      // When tests fail, save the logs.
+      protected void failed(Throwable t, Description description) {
+        try {
+          Files.copy(Paths.get(Constants.TESTS_LOG),
+              Paths.get(Constants.TEST_LOG_DIR, String.format("%s-%s.log",
+                  description.getClassName(), description.getMethodName())),
+              StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException e) {
+          if (t != null) {
+            t.addSuppressed(e);
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/BaseIntegrationTest.java]: Do we need to throw `t` here?

File: tests/src/test/java/alluxio/BaseIntegrationTest.java [hunk 3]
```diff
@@ -21,41 +23,65 @@
...
-                description.getClassName(), description.getMethodName())),
-            StandardCopyOption.REPLACE_EXISTING);
-      } catch (IOException e) {
-        if (t != null) {
-          t.addSuppressed(e);
-        } else {
-          throw new RuntimeException(e);
+  private RuleChain createRuleChain() {
+    RuleChain chain = RuleChain.emptyRuleChain();
+    chain = chain.around(logHandler());
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/BaseIntegrationTest.java]: Should we use `outerRule` here instead of `emptyRuleChain` + `around`?

**Comment by calvinjia** [comment]
@aaudiber Thanks for the proposal, I'm onboard with RuleChains.

File: tests/src/test/java/alluxio/BaseIntegrationTest.java (hunk 2 above)
Comment (aaudiber) [reply tests/src/test/java/alluxio/BaseIntegrationTest.java]: no, the base class handles exceptions and just notifies the subclass about them - that's why there's no `throws Exception` in the signature for `failed`

**Pull request merged**